Fix clippy issues

SomeoneToIgnore opened the pull request:
Fixes macOs clippy runs and moves clippy into GH Actions job.
The latter helps to simplify the clippy script and reduce overall CI run time.

hlinnaka commented:
If I understand correctly, with this patch, if you specify "--features profiling" on macOS it will nevertheless build without profiling support. Before this PR, you would get an error. That doesn't seem nice; if you ask for profiling, and profiling isn't supported on your platform, an error seems appropriate. Or how do rust crates usually behave with features that only work on some platforms?

SomeoneToIgnore commented:
Thank you for pointing out, I agree with the approach: whoever enables the feature on an incompatible platform earns the headache.

I was trying to fix two overlapping issues:
* `./run_clippy.sh` failing on macOs
* Basic `cargo clippy` fails on macOs

<details>
  <summary>The latter was caused by a few lints that I've fixed</summary>

```
warning: unneeded unit return type
  --> pageserver/src/profiling.rs:82:76
   |
82 |     pub fn profpoint_start(_conf: &PageServerConf, _point: ProfilingConfig) -> () {
   |                                                                            ^^^^^^ help: remove the `-> ()`
   |
   = note: `#[warn(clippy::unused_unit)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unused_unit

warning: unneeded unit expression
  --> pageserver/src/profiling.rs:83:9
   |
83 |         ()
   |         ^^ help: remove the final `()`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unused_unit

warning: unneeded unit return type
  --> pageserver/src/profiling.rs:86:48
   |
86 |     pub fn init_profiler(conf: &PageServerConf) -> () {
   |                                                ^^^^^^ help: remove the `-> ()`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unused_unit

warning: question mark operator is useless here
   --> pageserver/src/http/routes.rs:456:5
    |
456 |     Ok(json_response(StatusCode::OK, ())?)
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try removing question mark and `Ok()`: `json_response(StatusCode::OK, ())`
    |
    = note: `#[warn(clippy::needless_question_mark)]` on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_question_mark

error: calls to `std::mem::drop` with a value that implements `Copy`. Dropping a copy leaves the original intact
   --> pageserver/src/page_service.rs:391:13
    |
391 |             drop(profiling_guard);
    |             ^^^^^^^^^^^^^^^^^^^^^
    |
    = note: `#[deny(clippy::drop_copy)]` on by default
note: argument has type ()
   --> pageserver/src/page_service.rs:391:18
    |
391 |             drop(profiling_guard);
    |                  ^^^^^^^^^^^^^^^
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#drop_copy

warning: `pageserver` (lib) generated 4 warnings
error: could not compile `pageserver` due to previous error; 4 warnings emitted
```
<details>

The former was a bit harder, since we run clippy with `--all-features` in `./run_clippy.sh` script and that's a good thing to have even now.
I've reverted the platform-related conditional compilation and `Cargo.toml` conditional dependency and adjusted the CI config to run different clippy tasks based on the platform.

LizardWizzard reviewed (commented):
It looks like ./run_clippy.sh no longer runs in CI. It can be a reason for inconsistency (different options locally/remotely etc) Also script fails to run under mac os. I understand that it is not an issue introduced with this PR. What do you think about checking the OS in the script itself? Then script can be brought back to CI pipeline with no problems on mac os. Thoughts?

SomeoneToIgnore commented:
It is run in GitHub Actions CI, now in parallel to what's run in CircleCI: https://github.com/neondatabase/neon/pull/1568/files#diff-7dc87d4394e1756c519dbfd0b80d3b31377f643f0bc25d3ed807ce8a3794023dR67-R75

Linux is running the script: https://github.com/neondatabase/neon/runs/6141944526?check_suite_focus=true#step:10:1
Mac runs the "customized" version: https://github.com/neondatabase/neon/runs/6141944573?check_suite_focus=true#step:11:1

So, the script is in CI, just not for mac and that already has all of it minus the `profiling` feature, just written not in a form of a script.
Not 100% sure about adding the OS detection into the script right now, but got no strong arguments against it: I'm personally not fond of bash generally and not sure that attempts to detect in bash where it's run at worth for our case.
Since 99% of this is done, should we leave it for now before any need (new OS or new feature) to change this reappear?

LizardWizzard commented:
My concern is that the only needed difference is avoiding `--all-features` for mac os, but the script in CI for mac os has more deviations. It does not lint tests and missing `-A unknown_lints -D warnings`. It is a bit trickier to support, more places to make changes, etc.

Another option to avoid detection and use run_clippy on all platforms is to pass the platform as an argument. 

I'm ok with the current state, though I'll need to switch my local workflow to use manual clippy run with a custom command

SomeoneToIgnore commented:
Sure, have adjusted the script then.

LizardWizzard reviewed (approved):
Thanks!
